balancer/pickfirst/pickfirstleaf/pickfirstleaf_ext_test.go

```diff
@@ -851,6 +850,190 @@ func (s) TestPickFirstLeaf_EmptyAddressList(t *testing.T) {
 	}
 }
 
+// Test verifies that pickfirst correctly detects the end of the first happy
+// eyeballs pass when the timer causes pickfirst to reach the end of the address
+// list and failures are reported out of order.
+func (s) TestPickFirstLeaf_HappyEyeballs_TF_AfterEndOfList(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	originalTimer := pfinternal.TimeAfterFunc
+	defer func() {
+		pfinternal.TimeAfterFunc = originalTimer
+	}()
+	triggerTimer, timeAfter := mockTimer()
+	pfinternal.TimeAfterFunc = timeAfter
+
+	dialer := testutils.NewBlockingDialer()
+	opts := []grpc.DialOption{
+		grpc.WithDefaultServiceConfig(fmt.Sprintf(`{"loadBalancingConfig": [{"%s":{}}]}`, pickfirstleaf.Name)),
+		grpc.WithContextDialer(dialer.DialContext),
+	}
+	cc, rb, bm := setupPickFirstLeaf(t, 3, opts...)
+	addrs := bm.resolverAddrs()
+	holds := bm.holds(dialer)
+	rb.UpdateState(resolver.State{Addresses: addrs})
+	cc.Connect()
+
+	testutils.AwaitState(ctx, t, cc, connectivity.Connecting)
+
+	// Verify that only the first server is contacted.
+	if holds[0].Wait(ctx) != true {
+		t.Fatalf("Timeout waiting for server %d with address %q to be contacted", 0, addrs[0])
+	}
+	if holds[1].IsStarted() != false {
+		t.Fatalf("Server %d with address %q contacted unexpectedly", 1, addrs[1])
+	}
+	if holds[2].IsStarted() != false {
+		t.Fatalf("Server %d with address %q contacted unexpectedly", 2, addrs[2])
+	}
+
+	// Make the happy eyeballs timer fire once and verify that the
+	// second server is contacted, but the third isn't.
+	triggerTimer()
+	if holds[1].Wait(ctx) != true {
+		t.Fatalf("Timeout waiting for server %d with address %q to be contacted", 1, addrs[1])
+	}
+	if holds[2].IsStarted() != false {
+		t.Fatalf("Server %d with address %q contacted unexpectedly", 2, addrs[2])
+	}
+
+	// Make the happy eyeballs timer fire once more and verify that the
+	// third server is contacted.
+	triggerTimer()
+	if holds[2].Wait(ctx) != true {
+		t.Fatalf("Timeout waiting for server %d with address %q to be contacted", 2, addrs[2])
+	}
+
+	// First SubConn Fails.
+	holds[0].Fail(fmt.Errorf("test error"))
+
+	// No TF should be reported until the first pass is complete.
+	shortCtx, shortCancel := context.WithTimeout(ctx, defaultTestShortTimeout)
+	defer shortCancel()
+
+	testutils.AwaitNotState(shortCtx, t, cc, connectivity.TransientFailure)
+
+	// Third SubConn fails.
+	shortCancel()
```

Comment:
Do we need this? Won't `testutils.AwaitNotState` fail the test if the specified state is reached before the context expires?

Reply:
It's not required because of the way `testutils.AwaitNotState` works. When I tried to ignore the first cancel function as follows:
```go
shortCtx, _ := context.WithTimeout(ctx, defaultTestShortTimeout)
```
`govet` complains about a possible context leak because it can't ensure that the context will be cancelled at compile time. If we re-assign the `cancel` func later, `govet` doesn't complain but I still called `cancel` just to be consistent. Removed the call now.